### ISSUE TITLE
Rule proposal: `no-implicit-return-on-newline`

### ISSUE THREAD

**Issue by freesteph** [eslint/eslint#5989]
**When does this rule warn? Please describe and show example code:**

This rules checks that ES6 arrow functions that use implicit return do so on the same line as their parameters.

Correct:
```es6
// ex 1
someArray.map((value, index) => value + 1);

// ex 2
processAsync()
  .then(result => result
    .moreAsync()
    .then(final => Object.assign({}, final, { foo: 'bar' })));
```

Incorrect:
```es6
// ex 1
someArray.map((value, index) =>
  value + 1);

// ex 2
processAsync()
  .then(result =>
    result
      .moreAsync()
      .then(final =>
        Object.assign({}, final, { foo: 'bar' })));
``` 

**Is this rule preventing an error or is it stylistic?**

Purely stylistic. I find it much easier to process source-code if I can assume that same-line arrow-functions return and code blocks will explicitly `return`.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

This would benefit to any JS developers who also find it easier this way.

**Are you willing to create the rule yourself?**

Yes, if the rule makes sense to others.

**Comment by freesteph** [comment]
I'm also aware there has to be a better name for the rule.

**Comment by platinumazure** [comment]
Could you please refine your incorrect example so it would parse successfully? (You're missing a trailing parenthesis)

**Comment by freesteph** [comment]
whoops, amended.

**Comment by freesteph** [comment]
I've also added a slightly more complex example.

**Comment by platinumazure** [comment]
@freesteph Thanks for adding that second example. It does unfortunately raise a new question, though-- the outer arrow function in your second correct example spans more than one line, even though the inner one is only on one line. So this still needs to be fleshed out more. Why does the outer arrow function not warn in the second "correct" example?

**Comment by michaelficarra** [comment]
@platinumazure It appears to be because the expression *starts* on the same line. It doesn't matter whether it spans multiple lines or not. So this rule can be simplified to "don't put a newline directly following `=>`".

**Comment by freesteph** [comment]
I didn't realise it wasn't fully clear, my bad. So yes, the idea is that if you're using arrow functions' implicit returns then you need to "tie it" on the same line to clarify your intention.

I appreciate that an implicit return spanning multiple line:
```es6
.then(result => result
    .moreAsync()
```
might seem a bit odd at first, but I find it more explicit and less error-prone than constantly jumping back one line scanning for `{` to check whether the block is being returned or not.

**Comment by freesteph** [comment]
yes @michaelficarra, "don't put a newline directly following =>" works just as well here.

**Comment by nmn** [comment]
Name suggestion: `no-arrow-newline`

**Comment by mysticatea** [comment]
#6067 might be related.

**Comment by Trott** [comment]
I wonder if this might be better as an option on the existing `arrow-body-style` rule. Not sure, just pondering. If enabled and there is a newline after `=>` then the function body needs to be wrapped in `{...}`.

Good:

```js
callback((foo) => bar);

callback((foo) => {
  return bar;
});
```

Bad:

```js
callback((foo) =>
  bar);
```

**Comment by alberto** [comment]
@eslint/eslint-team does anyone think this should be in core and want to champion it?

**Comment by platinumazure** [comment]
I feel like this is the sort of thing that might have eventually found its way into JSCS, but I don't yet consider it critical for core.

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
